File: Data Structure/LinkedList/206.reverse_linked_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// Iterative solution
var reverseList = function(head) {
    if (head === null) return head;

    let prev = null;
    while (head != null) {
        let nextNode = head.next;
        head.next = prev;

        prev = head;
        head = nextNode;

    }

    return prev;
};

// Recursive solution -- 这方法不好做，尽量少用
var reverseList = function(head) {
    if (head === null || head.next === null) {
        return head;
    }
    let p = reverseList(head.next);
    head.next.next = head;
    head.next = null;
    return p;
}

File: Java/206.reverse_linkedlist.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
// Recursive
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode pre = null;
        ListNode cur = head;

        while (cur != null) {
            ListNode nextNode = cur.next;
            cur.next = pre;

            pre = cur;
            cur = nextNode;
        }

        return pre;
    }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        return reverseListHelper(head, null);
    }

    public ListNode reverseListHelper(ListNode head, ListNode pre) {
        if (head == null) return pre;

        ListNode nextNode = head.next;
        head.next = pre;
        return reverseList(nextNode, head);
    }
}